<a href="https://colab.research.google.com/github/longovinicius/Coursera_Capstone/blob/master/Third%20task.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Segmenting and Clustering Neighborhoods in **Toronto**

In this notebook, we will scrape a data frame from a wikipedia page. This data frame shows the Postal Code and Borough of each Neighborhood at the city of Toronto-Canada.


In [1]:
# Dependencies
import pandas as pd
import numpy as np


# FIRTS TASK

Using Pandas to scrape a dataframe


In [2]:
df = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')[0]
df.head()

Postal Code           Borough               Neighborhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

As we can see, there are some Postal Codes assigned for more than one neighborhood. In those cases, the neighborhoods are already at the same row, separated with coma.
We will drop the cells with assigned Boroughs, and we will replace unassigned neighborhoods with its Borough names.


In [3]:
# Drop unassigned Boroughs and replace unsigned Neighborhoods
df = df[df['Borough']!='Not assigned']
df['Neighborhood'].replace('Not assigned', df['Borough'], inplace=True)
df.head()

Postal Code           Borough                                 Neighborhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [4]:
df.reset_index()
df.shape

(103, 3)

The new data frame has 103 rows.

# SECOND TASK

We'll import a csv file containing every Postal Code from Toronto and its Latitude and Longitude
(geocoder wasn't working)

In [20]:
# getting latitude and longitude coordinates of each postal code from csv file

df_PC = pd.read_csv('http://cocl.us/Geospatial_data')
df_PC.head()


Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [21]:
# Merging both data frames by the 'Postal Code' column

df = pd.merge(df, df_PC, on='Postal Code')
df



Postal Code           Borough  ...   Latitude  Longitude
0           M3A        North York  ...  43.753259 -79.329656
1           M4A        North York  ...  43.725882 -79.315572
2           M5A  Downtown Toronto  ...  43.654260 -79.360636
3           M6A        North York  ...  43.718518 -79.464763
4           M7A  Downtown Toronto  ...  43.662301 -79.389494
..          ...               ...  ...        ...        ...
98          M8X         Etobicoke  ...  43.653654 -79.506944
99          M4Y  Downtown Toronto  ...  43.665860 -79.383160
100         M7Y      East Toronto  ...  43.662744 -79.321558
101         M8Y         Etobicoke  ...  43.636258 -79.498509
102         M8Z         Etobicoke  ...  43.628841 -79.520999

[103 rows x 5 columns]

# THIRD TASK

### Exploring

Getting the coordinates of Toronto using geopy

In [35]:
# dependencies
from geopy.geocoders import Nominatim
import folium
import json

address = 'Toronto, Ontario'

# getting geographic coordination of Toronto
geolocator = Nominatim(user_agent='to_explorer')
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('Geographical coordinate of Toronto are {}, {}'.format(latitude, longitude))

Geographical coordinate of Toronto are 43.6534817, -79.3839347


We will analyse the part of the city called 'Old Toronto'. In this part of the city, there are the boroughs: 'Downtown Toronto', 'East Toronto', 'West Toronto' and 'Central Toronto'. So we will create a new data frame only with the Boroughs with the word 'Toronto' in it.

In [36]:
# Creating a new data frame, only with Boroughs with the word 'Toronto' in it
df_new = df[df['Borough'].astype(str).str.contains('Toronto')]
df_new['Borough'].unique()

array(['Downtown Toronto', 'East Toronto', 'West Toronto',
       'Central Toronto'], dtype=object)

Let's create a map to visualize 'Old Toronto' and its neighborhoods

In [37]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=12)

for lat, lng, label in zip(df_new['Latitude'], df_new['Longitude'], df_new['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
map_toronto

Credentials of Foursquare (secret)

In [ ]:
CLIENT_ID = 'Foursquare ID' # Foursquare ID
CLIENT_SECRET = 'Foursquare Secret' # Foursquare Secret
VERSION = '20180605' # Foursquare API version


Funcition that returns the venues and its basic informations on each Neighborhood of 'Old Toronto'

In [39]:
def getNearbyVenues(names, latitudes, longitudes, radius=500, LIMIT=100):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Printing the data frame created by the function 'getNearbyVenues'

In [41]:
import requests

toronto_venues = getNearbyVenues(names=df_new['Neighborhood'],
                                   latitudes=df_new['Latitude'],
                                   longitudes=df_new['Longitude']
                                  )

print(toronto_venues.shape)
toronto_venues.head()

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West, Forest Hill Road Park
High Park, The Junction South
North Toronto West, Lawrence Park
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
R

Neighborhood  ...       Venue Category
0  Regent Park, Harbourfront  ...               Bakery
1  Regent Park, Harbourfront  ...          Coffee Shop
2  Regent Park, Harbourfront  ...  Distribution Center
3  Regent Park, Harbourfront  ...                  Spa
4  Regent Park, Harbourfront  ...           Restaurant

[5 rows x 7 columns]

### Analyse each neighborhood

Separating every category of the venues in a different column, creating a new data frame with 'dummy columns'

In [42]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Airport  ...  Wine Bar  Women's Store
0            0        0  ...         0              0
1            0        0  ...         0              0
2            0        0  ...         0              0
3            0        0  ...         0              0
4            0        0  ...         0              0

[5 rows x 226 columns]

grouping the 'toronto_onehot' data frame with each neighborhood

In [45]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()

toronto_grouped.head()

Neighborhood  ...  Women's Store
0                                        Berczy Park  ...            0.0
1       Brockton, Parkdale Village, Exhibition Place  ...            0.0
2  Business reply mail Processing Centre, South C...  ...            0.0
3  CN Tower, King and Spadina, Railway Lands, Har...  ...            0.0
4                                 Central Bay Street  ...            0.0

[5 rows x 226 columns]

In [46]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Creating new data frame with each neighborhood separated by most common category of venues 

In [47]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood  ... 10th Most Common Venue
0                                        Berczy Park  ...               Beer Bar
1       Brockton, Parkdale Village, Exhibition Place  ...             Restaurant
2  Business reply mail Processing Centre, South C...  ...             Restaurant
3  CN Tower, King and Spadina, Railway Lands, Har...  ...          Boat or Ferry
4                                 Central Bay Street  ...           Burger Joint

[5 rows x 11 columns]

### Clustering

Creating Kmeans object using only 2 k's as parameter and fitting it.

In [81]:
from sklearn.cluster import KMeans 
import matplotlib.cm as cm
import matplotlib.colors as colors

# set number of clusters
kclusters = 2

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
np.unique(kmeans.labels_, return_counts=True)

(array([0, 1], dtype=int32), array([33,  6]))

Merging a 'Cluster Labels' column, specifing what cluster the neighborhood was defined as

In [84]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df_new

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postal Code           Borough  ... 9th Most Common Venue  10th Most Common Venue
2          M5A  Downtown Toronto  ...        Ice Cream Shop                     Spa
4          M7A  Downtown Toronto  ...                  Café                Creperie
9          M5B  Downtown Toronto  ...             Bookstore       Electronics Store
15         M5C  Downtown Toronto  ...              Beer Bar     Moroccan Restaurant
19         M4E      East Toronto  ...            Donut Shop        Doner Restaurant

[5 rows x 16 columns]

Creating the final map with the clusters specified for each neighborhood

In [85]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Examine Clusters

Cluster 0

In [86]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  ...         10th Most Common Venue
2    Downtown Toronto  ...                            Spa
4    Downtown Toronto  ...                       Creperie
9    Downtown Toronto  ...              Electronics Store
15   Downtown Toronto  ...            Moroccan Restaurant
20   Downtown Toronto  ...                       Beer Bar
24   Downtown Toronto  ...                   Burger Joint
25   Downtown Toronto  ...                    Coffee Shop
30   Downtown Toronto  ...                      Bookstore
31       West Toronto  ...                        Brewery
36   Downtown Toronto  ...                          Plaza
37       West Toronto  ...                    Pizza Place
41       East Toronto  ...                         Lounge
42   Downtown Toronto  ...                       Beer Bar
43       West Toronto  ...                     Restaurant
47       East Toronto  ...                 Ice Cream Shop
48   Downtown Toronto  ...            Japanese Restaurant
54       East Toronto  ...                 Sandwich Place
67    Central Toronto  ...                     Donut Shop
69       West Toronto  ...                          Diner
73    Central Toronto  ...                           Park
74    Central Toronto  ...      Middle Eastern Restaurant
75       West Toronto  ...                   Dessert Shop
79    Central Toronto  ...             Seafood Restaurant
80   Downtown Toronto  ...             Italian Restaurant
81       West Toronto  ...                 Sandwich Place
84   Downtown Toronto  ...  Vegetarian / Vegan Restaurant
86    Central Toronto  ...                           Bank
87   Downtown Toronto  ...                  Boat or Ferry
92   Downtown Toronto  ...                       Pharmacy
96   Downtown Toronto  ...                 Sandwich Place
97   Downtown Toronto  ...             Seafood Restaurant
99   Downtown Toronto  ...                          Hotel
100      East Toronto  ...                     Restaurant

[33 rows x 12 columns]

Cluster 1

In [87]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  ...  10th Most Common Venue
19      East Toronto  ...        Doner Restaurant
61   Central Toronto  ...                 Dog Run
62   Central Toronto  ...        Doner Restaurant
68   Central Toronto  ...        Doner Restaurant
83   Central Toronto  ...              Donut Shop
91  Downtown Toronto  ...                 Dog Run

[6 rows x 12 columns]

As conclusion, we can see that Cluster 0 has mostly café's and restaurants as common venues, and we defined as Cluster 1 the neighborhoods with mostly Parks and Stores. 
Thank you for following! :) 